Introduce per-service-level workload types and their first use-case - shedding in interactive workloads

psarna opened the pull request:
This draft extends and obsoletes #8123 by introducing a way of determining the workload type from service level parameters, and then using this context to qualify requests for shedding.

The rough idea is that when the admission queue in the CQL server is hit, it might make more sense to start shedding surplus requests instead of accumulating them on the semaphore. The assumption that interactive workloads are more interested in the success rate of as many requests as possible, and hanging on a semaphore reduces the chances for a request to succeed. Thus, it may make sense to shed some requests to reduce the load on this coordinator and let the existing requests to finish.

It's a draft, because I only performed local guided tests. #8123 was followed by some experiments on a multinode cluster which I want to rerun first.

psarna commented:
... and I forgot to make it a github draft :( I'll follow by adding the `RFC` prefix to the title instead

psarna commented:
Idea: gather metrics for requests that *would* get shed, but the workload type is not set to be `interactive`, so the shedding does not happen; these metrics could help the users decide if they should declare some of their workloads as interactive to help reduce their load. The metrics should be labeled with their actual workload type (unspecified or batch).

avikivity commented:
Another thing we can do is enable empty pages in response to long sequences of tombstones for ETL workloads.

Although, we could do the same for non-ETL workloads. The client can easily choose to stop paging once the deadline is exceeded.

avikivity commented on service/qos/qos_common.cc:
```diff
@@ -56,7 +60,28 @@ service_level_options service_level_options::merge_with(const service_level_opti
             }
         },
     }, ret.timeout);
+    ret.workload = std::min(ret.workload, other.workload);
```
Does this not make unspecified override a specified workload?

avikivity replied:
I'd say interactive > batch > unspecified.

We should have thought of a way to prevent such conflicts, but too late now. Anyway, we just need to provide something, in practice it doesn't matter (though we must make specified override unspecified).

psarna replied:
The easiest thing to do would be to move `unspecified` as a last field in the enum and it would naturally work as expected, but I still prefer to have `unspecified` as 0, especially if we decide to add more workload types in the future. I'll just rewrite the conflict resolution logic in v2.

avikivity commented on db/system_distributed_keyspace.cc:
```diff
@@ -582,8 +584,10 @@ future<qos::service_levels_info> system_distributed_keyspace::get_service_levels
         qos::service_levels_info service_levels;
         for (auto &&row : *result_set) {
             auto service_level_name = row.get_as<sstring>("service_level");
+            auto workload = qos::service_level_options::parse_workload_type(row.get_opt<sstring>("workload_type").value_or(""));
```
This will throw if the value is wrong, killing the node. Better to ignore (with a warning) illegal values here.

psarna replied:
Also, two more issues regarding this piece of code:
 - killing the node is a preexisting condition, since we already parse service level name and the timeout duration in the same way
 - it's actually redundant to get the service level name from the row - it's already passed as a parameter to this function

I'll just add a commit which fixes both issues to this series

avikivity replied:
Doesn't parse_workload_type throw?

psarna replied:
Nope, it returns an optional. I guess I spent too much time with Rust to produce throwing code.

avikivity commented on transport/server.cc:
```diff
@@ -634,6 +655,12 @@ future<> cql_server::connection::process_request() {
+            if (_pending_requests_gate.get_count() == 0) {
+                _shed_incoming_requests = false;
+            }
+            return _read_buf.skip(length);
```
This is just to handle the timeout, yes? It should be mentioned and maybe even use handle_exception_type.

The extra continuation is unfortunate. Although, if there is no contention, it doesn't generate a task.

I hesitate about recommending to coroutinize this, since process_request()'s inner continuation (after read_frame()) will not wait in the common case.  We can't even fold it into the per-connection code since that's in some generic server.

psarna replied:
Its main purpose is for the timeout, but I intentionally put a generic `handle_exception` here, since *any* error (including some future ones) happening before the deferred action above (called `leave`) could leave the mechanism in a permanent shedding state - since there would be nobody to stop the process.

But, I now see that skipping `length` bytes in this handler is too eager - these bytes might have already been read by `read_and_decompress_frame`... I'd need to make this handler more robust.

psarna replied:
I'll solve this one by attaching the additional continuation right after getting semaphore units with a timeout. Additional bonus is that only the shedding path will now include this continuation.

avikivity commented on transport/server.cc:
```diff
@@ -577,6 +581,13 @@ future<> cql_server::connection::process_request() {
+        if (allow_shedding && _shed_incoming_requests) {
+            ++_server._stats.requests_shed;
+            return _read_buf.skip(f.length);
+        }
```
This is a little scary. We could spend many cycles looping and shedding requests, while those cycles can be used to retire existing requests. But the 50ms timer offers some protection, and serving a request is much more expensive than shedding. I think it's okay.

avikivity commented:
Very good!

psarna commented:
v2:
 - rewritten the conflict resolution rules in case of multiple applicable service levels: `batch` wins with `interactive`, both win with `unspecified`
 - added a cql-pytest case
 - added a unit test which checks conflict resolution (harder to do with cql-pytest because it implies authenticating with multiple roles and the harness is not quite ready for it yet)
 - added a paragraph in `docs/service_levels.md`
 - made getting service level information exception-safe
 - moved the continuation which catches semaphore errors earlier:
   - it now only allocates in the shedding path
   - it's more specific and only catches the expected semaphore error
   - it no longer introduces a bug which might have read from the buffer twice in rare cases

avikivity commented:
Dequeued, caused carnage in debug mode unit tests (presumably) and cql_query_test in release mode (bisected to commit c9d5845ff982f954f833f4281bfc4cc0f2a4ddc8
Author: Piotr Sarna <sarna@scylladb.com>
Date:   Wed May 19 10:34:21 2021 +0200

    cql3: add listing service level workload type
    
    The workload type information is now presented in the output
    of LIST SERVICE LEVEL and LIST ALL SERVICE LEVELS statements.
)

psarna commented:
I'll fix the issue and push v3 - it's a simple test issue and I think that I just forgot to recompile all unit tests before rerunning them

psarna commented:
And as for debug test carnage, it looks like a cross-core shared pointer usage was detected, I'd need to investigate

psarna commented:
Sneaky - in one place I used a static map of string -> data_type, but data_type is implemented as a singleton-like shared pointer, so unless the map is thread_local, all shards would use shard0's data type. Will push a fixed version in a minute

psarna commented:
v3:
 - fixed the outdated test case in cql_query_test
 - fixed a cross-shard issue with `data_type` - it was stored in a static non-thread-local array, which caused all shards to use shard0's shared pointers, which was thankfully caught by the debug build
